### PR TITLE
DiffractionFocussing2 bug with sorted events

### PR DESCRIPTION
**Pull request by peterfpeterson**
### Description of work

There was a difference in the result of `DiffractionFocussing` with `PreserverEvents` being true/false for sorted and unsorted events. The issue was finally discovered when creating the histogram view of a tof sorted event list (from compression) versus an unsorted event list (from file), which went through two different details for generating the histogram. For the tof sorted the Y-vector passed in is zeroed out. For the unsorted, the Y-vector is accumulated. This only occured for `PreserveEvents=False` in `DiffractionFocusing`

Fixes [EWM6757](https://ornlrse.clm.ibmcloud.com/ccm/web/projects/Neutron%20Data%20Project%20(Change%20Management)#action=com.ibm.team.workitem.viewWorkItem&id=6535). *There is no associated issue.*

### To test:

For people with access to the original 17GB data file this was discovered with
```python
from mantid.simpleapi import *


direc = '/absolute/path/to/data/'
wkspname='diamond'

# load data
Load(Filename=direc+'NOM_198202.nxs.h5', OutputWorkspace=wkspname, NumberOfBins=1)
LoadDiffCal(InputWorkspace=wkspname, Filename=direc+'NOMAD_196808_2023-08-14_shifter.h5', GroupFilename=direc+'NOM_Texture_Grouping.xml', WorkspaceName='NOM', TofMin=300)

# process up to focussing
MaskDetectors(Workspace=wkspname, MaskedWorkspace='NOM_mask')
CropWorkspace(InputWorkspace=wkspname, OutputWorkspace=wkspname, XMin=300)
CompressEvents(InputWorkspace=wkspname, OutputWorkspace=wkspname)
ResampleX(InputWorkspace=wkspname, OutputWorkspace=wkspname, NumberBins=6000, LogBinning=True)
ApplyDiffCal(InstrumentWorkspace=wkspname, CalibrationWorkspace='NOM_cal')
ConvertUnits(InputWorkspace=wkspname, OutputWorkspacediamond_hist=wkspname, Target='dSpacing')
ApplyDiffCal(InstrumentWorkspace=wkspname, ClearCalibration=True)

# EVENTS=FALSE
DiffractionFocussing(InputWorkspace=wkspname, OutputWorkspace=wkspname+"_hist", GroupingWorkspace='NOM_group', PreserveEvents=False)

#EVENTS=TRUE
DiffractionFocussing(InputWorkspace=wkspname, OutputWorkspace=wkspname, GroupingWorkspace='NOM_group', PreserveEvents=True)

# compare, ignoring that they are different types
CompareWorkspaces(Workspace1=wkspname+"_hist", Workspace2=wkspname, Checktype=False)
```

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

### CODE REVIEW

**Comment by Kvieta1990** [comment]
I think I am fine with this once the check passes. Thank you!

**Comment by Kvieta1990** [comment]
Grabbing the branch and tested it out with real NOMAD diamond data, I can confirm this is working! Thank you!

File: Framework/DataObjects/inc/MantidDataObjects/EventList.h [hunk 1]
```diff
@@ -267,9 +267,9 @@ class MANTID_DATAOBJECTS_DLL EventList : public Mantid::API::IEventList {
 
   void filterByPulseTime(Types::Core::DateAndTime start, Types::Core::DateAndTime stop, EventList &output) const;
 
-  void filterByPulseTime(Kernel::TimeROI *timeRoi, EventList *output) const;
+  void filterByPulseTime(Kernel::TimeROI const *timeRoi, EventList *output) const;
```
Comment (KyleQianliMa) [inline Framework/DataObjects/inc/MantidDataObjects/EventList.h:270]: Just a question, why add const in front of *timeRoi? In the future, is it better to define all *timeRoi as a const?

File: Framework/Algorithms/src/DiffractionFocussing2.cpp [hunk 2]
```diff
@@ -224,7 +224,16 @@ void DiffractionFocussing2::exec() {
       // get histogram version of the data
       if (eventInputWS) {
         const EventList &el = eventInputWS->getSpectrum(inWorkspaceIndex);
-        el.generateHistogram(group2xstep.at(group), Xout.rawData(), Yout, Eout);
+        // generateHistogram overwrites the data in Y and E so write to a temporary vector
+        MantidVec Ytemp;
+        MantidVec Etemp;
+        el.generateHistogram(group2xstep.at(group), Xout.rawData(), Ytemp, Etemp);
+        // accumulate the histogram into the output
+        std::transform(Ytemp.cbegin(), Ytemp.cend(), Yout.begin(), Yout.begin(),
```
Comment (KyleQianliMa) [inline Framework/Algorithms/src/DiffractionFocussing2.cpp:232]: I suppose this was causing issue before. Becaue Xout, Yout and Eout was overwritten. Now the data from generateHistorgram gets added into Yout and Eout.

**Review by KyleQianliMa — APPROVED** [review]
The code looks good to me. Just had a few minor questions.

File: Framework/DataObjects/inc/MantidDataObjects/EventList.h (hunk 1 above)
Comment (peterfpeterson) [reply Framework/DataObjects/inc/MantidDataObjects/EventList.h:270]: It was a suggestion made by cppcheck. `const Kernel::TimeROI *` and `Kernel::TimeROI const *` actually mean different things. [An article on the topic](https://www.geeksforgeeks.org/difference-between-constant-pointer-pointers-to-constant-and-constant-pointers-to-constants/).

**Pull request merged**